text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,236 @@
+(Additionally, the issue of dealing with nightly also makes crates.io hard to use, but that is a less clear-cut obstacle.)
+
+Cargo will soon gain the ability to create custom registries, as described in
+https://github.com/rust-lang/cargo/pull/2857 .
+Compilers should package the source of their implementation of the standard library as a registry, which can be distributed with the compiler.
+In practice, rustc will optionally contain the source in its sysroot.
+Rustup may be able to put it there if the default download does not contain it already.
+This registry will be called the "compiler source registry".
+
+## Cargo implementation
+
+Whenever Cargo encounters a `Cargo.toml`, the first thing it always does is inject any applicable implicit deps.
+The idea is that by doing this so early on, most of Cargo can stay the same in only knowing or caring about explicit deps, simplifying both this RFC and its implementation.
+This process is completely defined by the rules described in the first subsection, so there is really nothing to elaborate upon here.
+
+Cargo has a "source" abstraction for providers of packages.
+Examples of this are file-system paths, git repositories, and the upcoming registries.
+For the time being, `stdlib = true` will act like another source, instructing the compiler to look at the compiler source registry, and if that is missing, then the sysroot binaries.
+The remainder of this section will clarify this.
+
+The compiler source registry is expected to be located in `${$CARGO_RUSTC --print sysroot}/src`.
+The exact format this takes will be determined when custom registries are implemented.
+Based on that, it may be necessary to transform the rust repo before dumping it there.
+
+The "sysroot binary mock registry" is generated by examining the contents of the sysroot.
+Any binary in there will be added to the mock registry, with a version
+taken either from the `language-version` described above, or the compiler version if that key is not present (as it would be with existing rustc releases).
+
+If the compiler source registry exists, that is used, and the sysroot binary mock registry need not even be built.
+If the source registry is absent, then the binary registry is used.
+
+When the build plan just involves the compiler source registry and/or existing types of sources, it can be executed just like today.
+The awkward scenario is when packages from the sysroot binary mock registry need to be used in the build plan.
+Because Cargo doesn't know much about the sysroot binaries, it must be very conservative when deciding whether or not they can be used.
+For example, Cargo may assume they are built with only default features enabled but it can't know what those are.
+If features are explicitly requested, or the default features are disabled (by all dependent packages) then the binaries are ineligible for the build plan being resolved.
+Cargo likewise will have to be conservative with other metadata.
```

Comment:
Everything else that goes in a Cargo.toml, really. I just don't know off hand what matters for building. I think this is something that will only be clarified with implementation.